analyzers/tests/SonarAnalyzer.Test/Rules/UseConstantLoggingTemplateTest.cs

```diff
@@ -0,0 +1,158 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using CS = SonarAnalyzer.Rules.CSharp;
+
+namespace SonarAnalyzer.Test.Rules;
+
+[TestClass]
+public class UseConstantLoggingTemplateTest
+{
+    private readonly VerifierBuilder builder = CreateVerifier<CS.UseConstantLoggingTemplate>();
+
+    [TestMethod]
+    public void UseConstantLoggingTemplate_CS() =>
+        builder.AddPaths("UseConstantLoggingTemplate.cs").Verify();
+
+    [DataTestMethod]
+    [DataRow("Log", "LogLevel.Warning,")]
+    [DataRow("LogCritical")]
+    [DataRow("LogDebug")]
+    [DataRow("LogError")]
+    [DataRow("LogInformation")]
+    [DataRow("LogTrace")]
+    [DataRow("LogWarning")]
+    public void UseConstantLoggingTemplate_MicrosoftExtensionsLogging_CS(string methodName, string extraParameter = "") =>
+        builder.AddSnippet($$"""
+            using Microsoft.Extensions.Logging;
+
+            public class Program
+            {
+                public void Method(ILogger logger, int arg)
+                {
+                    logger.{{methodName}}({{extraParameter}} "Message");    // Compliant
+                    logger.{{methodName}}({{extraParameter}} $"{arg}");     // Noncompliant
+                }
+            }
+            """).Verify();
+
+    [DataTestMethod]
+    [DataRow("Debug")]
+    [DataRow("Error")]
+    [DataRow("Fatal")]
+    [DataRow("Info")]
+    [DataRow("Trace")]
+    [DataRow("Warn")]
+    public void UseConstantLoggingTemplate_CastleCoreLogging_CS(string methodName) =>
+        builder.AddSnippet($$"""
+            using Castle.Core.Logging;
+
+            public class Program
+            {
+                public void Method(ILogger logger, int arg)
+                {
+                    logger.{{methodName}}("Message");            // Compliant
+                    logger.{{methodName}}($"{arg}");             // Noncompliant
+                    logger.{{methodName}}Format("{Arg}", arg);   // Compliant
+                    logger.{{methodName}}Format($"{arg}");       // Noncompliant
+                }
+            }
+            """).Verify();
+
+    [DataTestMethod]
+    [DataRow("Debug")]
+    [DataRow("Error")]
+    [DataRow("Fatal")]
+    [DataRow("Info")]
+    [DataRow("Warn")]
+    public void UseConstantLoggingTemplate_Log4Net_CS(string methodName) =>
+        builder.AddSnippet($$"""
+            using log4net;
+
+            public class Program
+            {
+                public void Method(ILog logger, int arg)
+                {
+                    logger.{{methodName}}("Message");               // Compliant
+                    logger.{{methodName}}($"{arg}");                // Noncompliant
+                    logger.{{methodName}}Format("Arg: {0}", arg);   // Compliant
+                    logger.{{methodName}}Format($"{arg}");          // Noncompliant
```

Comment:
The rule will check whether the method has the same name, and an argument with the given name (message, format, etc.) to match the invocation. So the rule supports the overloads you listed.
I've added [UTs to test this Microsoft.Extensions.Logging.ILogger](https://github.com/SonarSource/sonar-dotnet/blob/fcf5bde34d818fa79187a9e76babaafeae66a0bc/analyzers/tests/SonarAnalyzer.Test/TestCases/UseConstantLoggingTemplate.cs#L40), but I'd rather not do it for other methods in frameworks, because for 5 frameworks with roughly 6-7 methods, each having multiple overloads would be massive.